### DISCUSSION TITLE
use of $primary instead of $theme-color("primary") in _variables.scss complicates color overrides

### DISCUSSION THREAD

**Discussion by bellenss** [twbs/bootstrap#32651] [General] [answered]
Whereas in v4.* all variables used the $theme-color("primar") syntax, the v5.0.0-beta1 version utilizes $primary syntax. 
In case of color customization, this means that overriding the $theme-colors map is insufficient. One also has to override each individual color variable ($primary, $secondary, ...).
What is the reason of this change from $theme-color("primary") to $primary in the _variables.scss definition?

**Comment by ffoodd** [comment, marked as answer]
Mainly because `theme-color()` only did a `map-get()`, which in turns only called a variable —all of this explained in #29083. Calling the variable directly reduces cyclomatic complexity (thus lowers compilation time by being more performant, saves CPU thus energy, etc.).

Another reason (resulting from the previous point) is to ensure to have a value: since `$theme-colors` is very likely to be overriden, the `primary` key might be dropped, resulting in `theme-color('primary')` being `null`. By using `$primary`, we ensure to have a value and encourage variable overrides instead of map ones. The same reasons explain why we're using `$spacer / 2` instead of `map-get($spacers, 2)` for example.

Hope that helps :)

**Comment by MonsoonBloom** [comment]
It is SOOOOOOO complicated to change the theme colors. Utilities to compile do not work. Code does not work. No one has the correct solution. Why in the world would we make changing theme colors so obtuse? Instructions on the Bootstrap Theming page documentation is brief and does not describe anything. Literally, I have been reading documentation, watching videos, installing and uninstalling packages that do not work, testing possible solutions for 6 hours now. This is complete nonsense and it makes  BOOTSTRAP UNUSABLE. The whole point was to be able to change stuff at the top level so one would not have to go in and write miles of css code. now with sass, that is nearly (actually) IMPOSSIBLE.

**Comment by Dacesilian** [comment]
Can someone please tell me, how can I override color? I've used code from here (https://getbootstrap.com/docs/5.0/customize/sass/):

custom.scss:
```
$green: #28a745;
@import "bootstrap";
```

File is placed in scss folder and then `npm run dist`. But color is not changed in result.
Thanks.

**Reply by bellenss** [reply]
https://blog.assistancy.be/blog/bootstrap-5-override-theme-colors-sass

**Reply by Dacesilian** [reply]
> https://blog.assistancy.be/blog/bootstrap-5-override-theme-colors-sass

I've read that, but I'm really not sure how to use it - they override $success, but I want to override base color ($green). Why isn't there any clear tutorial? :(

**Reply by ffoodd** [reply]
What's confusing here is that overriding `$green` won't override automagically `$success` or `$theme-colors` map. To effectively apply your `$green` change you need to override each variable using it (eg. `$success`, `$theme-colos`, etc.).

**Reply by bellenss** [reply]
This is actually not right. I made a small test and confirm that the desired behaviour is working at my side. This is with Bootstrap 5.0.1. 
Define any custom green color, and load Bootstrap afterwards. all variables (e.g. success color) use the new green color.

Two remarks:
* are you recompiling bootstrap after any *.scss change? Bootstrap won't pick up the change automatically, you need to recompile it yourself using the Sass precompiler. https://getbootstrap.com/docs/5.0/getting-started/build-tools/#sass
* I would ask the question 'why override a default $green variable? As the name suggests, it's a color definition. I believe it makes more sense in this case to override the resulting output variable, e.g. $success in the theme color map, as it makes semantically more sense.
